### DISCUSSION TITLE
Redash installation script now supports many Linux distros

### DISCUSSION THREAD

**Discussion by justinclift** [getredash/redash#6830] [Announcements]
The official Redash installation script ([setup.sh](https://github.com/getredash/setup)) has until now only worked on Ubuntu Linux.

That has now changed, and as of today it now happily runs on several more (all tested):

- Alma Linux 8.x & 9.x
- CentOS Stream 9.x
- Debian 12.x
- Fedora 38, 39 & 40
- Oracle Linux 9.x
- Red Hat Enterprise Linux 8.x & 9.x
- Rocky Linux 8.x & 9.x
- Ubuntu LTS 20.04 & 22.04

If anyone wants to add support for more, or even add support for other platforms (macOS?, WSL on Windows?) then they'd be welcome contributions. :smile:

---

It's recently had a new `--preview` option added too, for installing the Redash `preview` image rather than the (now very old) highest official release.  Use it like this:

```
$ sudo ./setup.sh --preview
```

**Comment by justinclift** [comment]
Note, that's x86_64 only (so far).  ARM64 support will happen at some point too, but it's not there right now. :wink:

**Comment by ahmgithubahm** [comment]
That's great news!

As you'll have seen from my very recent questions here, I used the previous setup.sh as a guide to install on Debian 12.  Is there an easy upgrade path or guide, to go from redash 8.0.x to 10.1.x, and also the fact that PostgreSQL looks like it will get upgraded from 15 to 16, particularly with respect to what changes I should make to my docker compose file?

thanks

**Reply by justinclift** [reply]
@ahmgithubahm Well, first things first... make a backup of your database and the environment file (likely `/opt/redash/env`) that matches it before doing anything.  Really not joking. :smile:

> Is there an easy upgrade path or guide ...

Yeah, there's an official [Upgrade Guide](https://redash.io/help/open-source/admin-guide/how-to-upgrade).  The steps there are right, though the versions listed are too out of date and should be updated.

Even the most recent official release tag (`10.1.0.b50633`) is pretty old and can't connect to some kind of data sources (older MongoDB comes to mind).  The `preview` image is automatically built from development code though, so can be hit or miss.

We *really* need to sort out a new official release that's production quality, but things aren't there yet.

---

For the PostgreSQL version change though... it's kind of optional.  Redash doesn't actually make use of any of the newer PostgreSQL features.  It's just that the old version of PostgreSQL which Redash uses is outside of the PostgreSQL project's "supported list".  So we've moved things to a newer version, and it's one that will automatically upgrade the database software itself without user intervention.  In theory. :pray:

---

Does that help? :smile:

**Reply by ahmgithubahm** [reply]
Thanks for the speedy reply!

I assume your (sensible!) database backup suggestion refers to the docker container PostgreSQL database, rather than the external data sources ?  I'm not that familiar with running databases in containers (all our prod stuff is AWS Aurora), but is it enough to just backup the subdirectory /opt/redash/postgres-data that is bind-mounted into the redash-postgres container (while it's all stopped obviously)?  \<edit> I see the upgrade steps do clarify that it's the PostgreSQL in the container.

I take your point about redash not needing the latest PG, but since I only just set up my redash service, this is the best time to take on board these changes.

Thanks for the pointer to the upgrade steps.  It mentions going one version at a time, but since there was only ever a 9.x beta, is it OK in this case to go straight from 8.0.x to 10.1.x?  Certainly the release notes (https://github.com/getredash/redash/releases) seem to imply that is fine.

Thank you

**Reply by ahmgithubahm** [reply]
Decided to proceed anyway, since I'm at the 'empty redash' stage so no harm if it crashes and burns.

I started by moving to the latest pgautoupgrade (PostgreSQL) image, which carried out the 15->16 DB upgrade fine.

Following the steps under https://github.com/getredash/redash/releases/tag/v10.0.0, I then moved to redash 10.0.0.b50363, also applying the change to the compose file for the stuff related to the worker.  This went fine, although I think the `docker-compose up --force-recreate --build` command either needs a `-d` or an instruction to hit Ctrl-C after it's finished doing stuff.

I could have gone straight to 10.1 but I got carried away following the 10.0 instructions exactly, and forgot there is a 10.1.  Anyway, I then moved from 10.0 to 10.1 fine.

Everything looks good, and I'm at redash 10.1 and PG 16 now.

Thanks!

**Reply by justinclift** [reply]
>  This went fine, although I think the docker-compose up --force-recreate --build command either needs a -d or an instruction to hit Ctrl-C after it's finished doing stuff.

Oops.  Yeah, that needs a `-d` on the end. :smile:

And seems like it's all gone well.  Cool. :grin:

**Reply by justinclift** [reply]
> I assume your (sensible!) database backup suggestion refers to the docker container PostgreSQL database, rather than the external data sources ?

Yep, that's correct.  It's just the docker container PostgreSQL database that needs backing up, as that's what the upgrade process changes.  And yep, the super foolproof way of backing that up is to just copy the `/opt/redash/postgres-data` directory somewhere while Redash containers are shut down.

Note - There are ways to back it up (eg `pg_dump`) while everything is still running too.  It's a bit more of a pain to do if you're not familiar with things though. :smile:

**Reply by JiffsMaverick** [reply]
@ahmgithubahm Hello! I am currently using Redash version 10.1.0 with PostgreSQL 9.6. I would like to upgrade the database to version 16. While researching how to do this, I came across this discussion.

I see that you managed to upgrade the database. I would like to clarify if there were any issues related to Redash dependencies. Based on this [discussion](https://github.com/getredash/redash/discussions/5933), it seems there is an issue with the SQLAlchemy-Searchable library, which only supports older versions of PostgreSQL. There is even an explanation about it [here](https://github.com/getredash/redash/blob/v10.1.0/requirements.txt):

```
# We can't upgrade SQLAlchemy-Searchable version as newer versions require PostgreSQL > 9.6, but we target older versions at the moment.
SQLAlchemy-Searchable==0.10.6
```

Did you encounter any problems with this on Redash version 10.1.0?

**Reply by justinclift** [reply]
@JiffsMaverick In theory (!), you should be able to upgrade the PostgreSQL version by just changing the name of the PostgreSQL Docker image to `pgautoupgrade/pgautoupgrade:latest` then restarting the containers.

Like this:

https://github.com/getredash/setup/blob/1299b8fb3e68b19890e736785a9d1e6a76a3d13c/data/compose.yaml#L41

Before you do that though, make sure you backup the database + matching environment file (`/opt/redash/env`).  Just in case. :smile:

**Reply by JiffsMaverick** [reply]
@justinclift Thank you for your reply. Yes, I saw your commit. But I thought it only works in conjunction with the new version of SQLAlchemy-Searchable, which is already in the [master branch](https://github.com/getredash/redash/blob/master/pyproject.toml#L74).

Do you have any working examples of Redash version 10.1.0 (with the old version of the SQLAlchemy-Searchable library) working with PostgreSQL 16?

**Reply by justinclift** [reply]
> Do you have any working examples of Redash version 10.1.0 (with the old version of the SQLAlchemy-Searchable library) working with PostgreSQL 16?

We used to run production instances of Redash 10.1.0 with PostgreSQL 13 & 15, so those are known to be good in real world use.  I'm not super sure about PostgreSQL 16, though like with the previous versions it *should* (in theory) be fine as Redash doesn't use any advanced features.

You can't just blindly upgrade between (standard) PostgreSQL versions without effort though.  That's what the `pgautoupgrade` Docker containers automate, but you can absolutely do the same thing manually if that's more acceptable for some reason.

Alternatively, it's also possible to upgrade the PostgreSQL database by exporting all of the PostgreSQL data (using `pg_dumpall`), then starting a fresh new instance of the newer PostgreSQL and loading the data into that.

---

That being said, ***really*** do make sure you have a backup of the database and matching environment file (contains encryption keys) before attempting things.

With such a backup, you could load that into a new "test" instance of Redash on another computer using the newer PostgreSQL just to check it works.  That would be a fairly standard way to verify things for yourselves without impacting your proper Redash installation.

---

Does that help? :smile:

**Comment by lucydodo** [comment]
https://github.com/getredash/setup/pull/64: Fedora distributions now also support install via script.
